#2529 Adds a cli flag to skip checks at startup.

mkutsevol opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
There are use cases when ingress is deployed with host networking and 80/443 ports are already bound on some ips, but not on the ip dedicated to ingress.
This allows for a cheap 'offload to user' workaround for that case.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2529

**Special notes for your reviewer**:

antoineco commented:
Maybe we should have a flag to configure the listen address instead, and check ports availability accordingly?

mkutsevol commented:
I haven't gone through all the code/use cases, _maybe_ there is a use case for dynamic change of the binding address. I suspect something from loadbalancer's world. 
I mean that bind address is configurable in configmap.

JordanP commented:
Yeah, I'd rather see a `--listen-address` flag. Not sure anyone would need a dynamic listen address anytime soon, let's start simple.

mkutsevol commented:
I mean it's already there. We have a configmap. The changes are watched & reapplied, so changing https://github.com/kubernetes/ingress-nginx/blob/326ceec3c756e23a2a7b3aba6934fb0fab669cff/docs/user-guide/nginx-configuration/configmap.md#bind-address 
in the configmap will change bind address. 
I haven't checked that  manually, but from the first glance it should be that way.

aledbf commented:
@mkutsevol you are right, this is being configured in the configmap and not in the flags. This was a mistake because we should not start the controller if we cannot bind to the specified IP address. 
If you have the time, please remove the bind-address code from the configmap and create a flag and a field [here](https://github.com/kubernetes/ingress-nginx/blob/master/cmd/nginx/flags.go#L226)

mkutsevol commented:
Ok, sounds good, I'll add --bind-address *flag* and test coverage, will do it next Sunday I think.

mkutsevol commented:
Didn't make it to test coverage, will have more time a bit later. 
Implemented the flag and cleaned up the default nginx.conf template. 
I like it better this way, less repetitions and should be easier to understand by a newcomer.
Please look at the current implementation and if it's ok, I'll proceed with tests as I have time. 
Cheers!

aledbf commented on cmd/nginx/flags.go:
```diff
@@ -147,6 +149,8 @@ Requires the update-status parameter.`)
 			`Dynamically refresh backends on topology changes instead of reloading NGINX.
 Feature backed by OpenResty Lua libraries.`)
 
+		bindAddresses = flags.String("bind-addresses", "0.0.0.0,[::]",
```
This should be a valid IPV6 address (remove the [] chars)

antoineco replied:
No need to add "Default..." in the flag description. The flag package does it for you during compilation.

aledbf commented:
@mkutsevol this looks a lot cleaner. Besides the tests this lgtm. Thanks for the cleanup!

antoineco commented on cmd/nginx/flags.go:
```diff
@@ -188,25 +192,47 @@ Feature backed by OpenResty Lua libraries.`)
-	}
+	bindIpAddresses := make([]string, 0)
+	for _, ipAddressStr := range strings.Split(*bindAddresses, ",") {
+		ipAddressSanitized := strings.Replace(strings.Replace(ipAddressStr, "]", "", -1),
```
The strings.Trim... functions are more suitable for this. Besides, do we need brackets at all? (cf. previous @aledbf's comment)

antoineco commented on cmd/nginx/flags.go:
```diff
@@ -188,25 +192,47 @@ Feature backed by OpenResty Lua libraries.`)
+	// check port collisions
+	for _, ipAddress := range bindIpAddresses {
+		if !ing_net.IsPortAvailable(ipAddress, *httpPort) {
+			return false, nil, fmt.Errorf("port %v:%v is already in use. Please check the flag --http-port",
```
(nit) Printed errors should start with a capital letter, and you could use %d to make it clear you're printing an integer.

mkutsevol commented:
Addressed the comments, thanks  for looking into it. :)
Will proceed with tests as soon as I have time.

dohnto commented:
This is great, this issue is blocking us from using ingress-nginx. Can I help with something to get this merged?

mkutsevol commented:
@dohnto the tests are broken, feel free to pick up and fix broken e2e tests and add coverage for the new code. 
I'm trying to allocate some time for it next week.

aledbf commented:
Closing. Please reopen when you can update the PR and pass CI
